ouroboros-consensus-cardano/src/ouroboros-consensus-cardano/Ouroboros/Consensus/Cardano/QueryHF.hs

```diff
@@ -0,0 +1,126 @@
+{-# LANGUAGE DataKinds           #-}
+{-# LANGUAGE EmptyCase           #-}
+{-# LANGUAGE FlexibleContexts    #-}
+{-# LANGUAGE FlexibleInstances   #-}
+{-# LANGUAGE LambdaCase          #-}
+{-# LANGUAGE MonoLocalBinds      #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+{-# OPTIONS_GHC -Wno-orphans #-}
+
+module Ouroboros.Consensus.Cardano.QueryHF () where
+
+import           Data.SOP.Index
+import           Data.SOP.Strict
+import           Ouroboros.Consensus.Byron.Ledger
+import           Ouroboros.Consensus.Cardano.Block
+import           Ouroboros.Consensus.Cardano.CanHardFork
+import           Ouroboros.Consensus.Cardano.CanonicalTxIn ()
+import           Ouroboros.Consensus.HardFork.Combinator.Ledger.Query
+import           Ouroboros.Consensus.Ledger.Query
+import           Ouroboros.Consensus.Shelley.Ledger.Query
+import           Ouroboros.Consensus.TypeFamilyWrappers
+
+instance CardanoHardForkConstraints c => QueryLedgerHF (CardanoEras c) where
```

Comment:
@jasagredo and I chatted. We think something like this might be possible:

```answerBlockQueryHFOne idx@(IS IZ) cfg q dlv =
    cardanoFoldSum idx (Proxy @AnswerShelleyOne)
  $ FoldSumHandlers {handleByron = \Refl -> case q of {}, handleShelley = \Refl -> answerShelleyOne idx cfg q dlv}
```

Which could hide the SoP implementation details in the (many?) places like this that have all three of these:

- the concrete `CardanoBlock` context
- one implementation for `ByronBlock`
- one uniform implementation for every `ShelleyBlock`